File: Java Code/src/code/ShopOwner.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.mycompany.rts_assignment_final;

import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.TimeoutException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Dell
 */
public class ShopOwner implements Runnable{
    
    static Flour flour;
    static Residue residue;
    static Milk milk;
    
    static ArrayList<Flour> flourRack = new ArrayList<>();
    static ArrayList<Residue> residueRack = new ArrayList<>();
    static ArrayList<Milk> milkRack = new ArrayList<>();
    
    static int counter = 0;
    static int type = 0; //1.flour 2.milk 3.residue

    @Override
    public void run() {
        
        try {
            String exchange = "Rabbit MQ Customer Topic";
            
            String key_FnR = "FLOUR.RESIDUE";
            String key_FnM = "MILK.FLOUR";
            String key_RnM = "RESIDUE.MILK";
            String key_FnRnM = "FLOUR.RESIDUE.MILK";
            
            byte[] FlourByteArray = Serialization.getByteArray2(flour);
            byte[] ResidueByteArray = Serialization.getByteArray3(residue);
            byte[] MilkByteArray = Serialization.getByteArray4(milk);
            
            System.out.println("SHOP OWNER: Serving Customer " + Customer.ID_List.get(0) + "...");
            
            if("*.FLOUR".equals(Customer.bindindKey)){
                type = 1;
                RabbitMQ.producer(exchange, FlourByteArray, key_FnM, "t");
                type = 2;
                RabbitMQ.producer(exchange, MilkByteArray, key_FnM, "t");
            }
            else if("FLOUR.*".equals(Customer.bindindKey)){
                type = 1;
                RabbitMQ.producer(exchange, FlourByteArray, key_FnR, "t");
                type = 3;
                RabbitMQ.producer(exchange, ResidueByteArray, key_FnR, "t");
            }
            else if("RESIDUE.*".equals(Customer.bindindKey)){
                System.out.print        ("");
                type = 3;
                RabbitMQ.producer(exchange, ResidueByteArray, key_RnM, "t");
                type = 2;
                RabbitMQ.producer(exchange, MilkByteArray, key_RnM, "t");
            }
            else{
                type = 1;
                RabbitMQ.producer(exchange, FlourByteArray, key_FnRnM, "t");
                type = 3;
                RabbitMQ.producer(exchange, ResidueByteArray, key_FnRnM, "t");
                type = 2;
                RabbitMQ.producer(exchange, MilkByteArray, key_FnRnM, "t");
            }
            
        } catch (IOException ex) {
            Logger.getLogger(ShopOwner.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    public static void collect(){
        
        String exchange = "Soybean Direct";
        
        String key1 = "Flour Key";
        String key2 = "Residue Key";
        String key3 = "Milk Key";
        
        try {
            RabbitMQ.consumer(exchange, key1, "d");
            RabbitMQ.consumer(exchange, key2, "d");
            RabbitMQ.consumer(exchange, key3, "d");
            
        } catch (IOException ex) {
            Logger.getLogger(Worker.class.getName()).log(Level.SEVERE, null, ex);
        } catch (TimeoutException ex) {
            Logger.getLogger(Worker.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    public static void collectFlour() throws IOException, ClassNotFoundException{
        
        counter++;
        
        byte[] object = RabbitMQ.return_object();
        
        flour = (Flour) Serialization.deserialize(object);
        
        flourRack.add(flour);
        
    }
    
    public static void collectResidue() throws IOException, ClassNotFoundException{
        
        counter++;
        
        byte[] object = RabbitMQ.return_object();
        
        residue = (Residue) Serialization.deserialize(object);
        
        residueRack.add(residue);
        
    }
    
    public static void collectMilk() throws IOException, ClassNotFoundException{
        
        counter++;
        
        byte[] object = RabbitMQ.return_object();
        
        milk = (Milk) Serialization.deserialize(object);
        
        milkRack.add(milk);
        
        if(counter == 45){
            System.out.println("\nSHOPOWNER: Purchased " + flour + " & " + residue + " & " + milk + " from the WORKER");
            counter = 0;
        }
    }
}

File: Rust Code/src/ShopOwner.rs
extern crate arraylist;

use crate::{Amiquip, Flour, Milk, Residue};
use std::cell::RefCell;
use std::fs::copy;
use std::time::Duration;

use arraylist::arl::ArrayList;
use arraylist::arraylist;

pub struct shopOwner;

/*pub struct flourRack{
    vec: RefCell<Vec<Flour>>,
}*/


pub static mut product_Type: i32 = 0;

impl shopOwner{

    pub fn run(&self, mut bindingKey:String, mut customerID:i32){

        let key_FnR = "FLOUR.RESIDUE";
        let key_FnM = "MILK.FLOUR";
        let key_RnM = "RESIDUE.MILK";
        let key_FnRnM = "FLOUR.RESIDUE.MILK";

        let exchange_name = "Rabbit MQ Customer Topic";

        let mut flour = Flour{name: String::from("flour")};
        let mut residue = Residue{name: String::from("residue")};
        let mut milk = Milk{name: String::from("milk")};

        let FlourBytes = bincode::serialize(&flour).unwrap();
        let ResidueBytes = bincode::serialize(&residue).unwrap();
        let milkBytes = bincode::serialize(&milk).unwrap();

        //std::thread::sleep(Duration::from_secs(1));
        //println!("SHOP OWNER: Serving Customer {} ...", customerID);
        if(bindingKey == "*.FLOUR"){
            unsafe { product_Type = 1; }
            Amiquip::producer(exchange_name,FlourBytes,key_FnM,"t");
            //println!("SHOP OWNER: Serving Customer {}...", customerID);
            unsafe { product_Type = 2; }
            Amiquip::producer(exchange_name,milkBytes,key_FnM,"t");
        }
        else if (bindingKey == "FLOUR.*") {
            unsafe { product_Type = 1; }
            Amiquip::producer(exchange_name,FlourBytes,key_FnR,"t");
            unsafe { product_Type = 3; }
            Amiquip::producer(exchange_name,ResidueBytes,key_FnR,"t");
        }
        else if (bindingKey == "RESIDUE.*") {
            unsafe { product_Type = 3; }
            Amiquip::producer(exchange_name,ResidueBytes,key_RnM,"t");
            unsafe { product_Type = 2; }
            Amiquip::producer(exchange_name,milkBytes,key_RnM,"t");
        }
        else {
            unsafe { product_Type = 1; }
            Amiquip::producer(exchange_name,FlourBytes,key_FnRnM,"t");
            unsafe { product_Type = 3; }
            Amiquip::producer(exchange_name,ResidueBytes,key_FnRnM,"t");
            unsafe { product_Type = 2; }
            Amiquip::producer(exchange_name,milkBytes,key_FnRnM,"t");
        }



    }

    pub fn collect(&self){

        let exchange_name = "Soybean Direct";

        let key1 = "Flour key";
        let key2 = "Residue Key";
        let key3 = "Milk key";

        Amiquip::consumer(exchange_name,key1,"d");
        Amiquip::consumer(exchange_name,key2,"d");
        Amiquip::consumer(exchange_name,key3,"d");
    }



    pub fn collectFlour(mut flour:Flour){
        println!("SHOP OWNER: {} is received!", flour.to_string());
    }

    pub fn collectResidue(mut residue:Residue){
        println!("SHOP OWNER: {} is received!", residue.to_string());
    }

    pub fn collectMilk(mut milk:Milk){

        println!("SHOP OWNER: {} is received!", milk.to_string());
    }
}

/*impl flourRack {
    pub fn created()->flourRack {
        flourRack {vec: RefCell::new(Vec::new()) }
    }
    pub fn add(&self, value: Flour){
        self.vec.borrow_mut().push(value);
        println!("{:?}", self.vec);
    }
}*/


